File: frontend/src/lib/api/story.ts
import { queryOptions } from "@tanstack/react-query";
import { createServerFn } from "@tanstack/react-start";
import { backendFetch } from "#/lib/fetch";
import type { StoryArchive } from "#/types/generated/StoryArchive";
import type { StoryCommunity } from "#/types/generated/StoryCommunity";
import type { StoryIllustrations } from "#/types/generated/StoryIllustrations";
import type { StoryIndex } from "#/types/generated/StoryIndex";
import type { StoryScript } from "#/types/generated/StoryScript";
import { DEFAULT_GAMEDATA_SERVER, gamedataPath, resolveGamedataServer } from "./gamedata";

/** The Archives library: every story group and every operator's records. See `docs/story-reader.md`. */
export const getStoryIndexFn = createServerFn({ method: "GET" })
    .inputValidator((server: string) => server)
    .handler(async ({ data: server }) => {
        const res = await backendFetch(gamedataPath(server, "/story/index"));
        if (!res.ok) throw new Error(`Failed to load story index: ${res.status}`);
        return (await res.json()) as StoryIndex;
    });

export function storyIndexQueryOptions(server: string = DEFAULT_GAMEDATA_SERVER) {
    return queryOptions({
        queryKey: ["story", "index", resolveGamedataServer(server)],
        queryFn: () => getStoryIndexFn({ data: resolveGamedataServer(server) }),
        staleTime: 60 * 60 * 1000,
        gcTime: 24 * 60 * 60 * 1000,
    });
}

/** One parsed script, or `null` when the library lists the id but no script file backs it (404). */
export const getStoryFn = createServerFn({ method: "GET" })
    .inputValidator((data: { id: string; server: string }) => data)
    .handler(async ({ data: { id, server } }) => {
        const res = await backendFetch(gamedataPath(server, `/story/${encodeURIComponent(id)}`));
        if (res.status === 404) return null;
        if (!res.ok) throw new Error(`Failed to load story: ${res.status}`);
        return (await res.json()) as StoryScript;
    });

export function storyQueryOptions(id: string, server: string = DEFAULT_GAMEDATA_SERVER) {
    return queryOptions({
        queryKey: ["story", "script", resolveGamedataServer(server), id],
        queryFn: () => getStoryFn({ data: { id, server: resolveGamedataServer(server) } }),
        staleTime: 60 * 60 * 1000,
        gcTime: 24 * 60 * 60 * 1000,
    });
}

/**
 * One group's illustrations, or one operator's: the backend takes a
 * `StoryGroup.id` and an operator `charId` on the SAME route, because an
 * operator's records are split over one or two story sets and have no single
 * group id of their own. The shape is the generated binding, fixed in
 * `docs/story-reader.md` section 2.
 */
/**
 * A group's illustrations, or `null` when the running backend does not serve
 * the route (404). Null is a first-class answer here: the endpoint shipped
 * after this tab did, and a page that threw would take the whole library down
 * with it until a restart.
 */
export const getStoryIllustrationsFn = createServerFn({ method: "GET" })
    .inputValidator((data: { groupId: string; server: string }) => data)
    .handler(async ({ data: { groupId, server } }) => {
        const res = await backendFetch(gamedataPath(server, `/story/group/${encodeURIComponent(groupId)}/illustrations`));
        if (res.status === 404) return null;
        if (!res.ok) throw new Error(`Failed to load illustrations: ${res.status}`);
        return (await res.json()) as StoryIllustrations;
    });

export function storyIllustrationsQueryOptions(groupId: string, server: string = DEFAULT_GAMEDATA_SERVER) {
    return queryOptions({
        queryKey: ["story", "illustrations", resolveGamedataServer(server), groupId],
        queryFn: () => getStoryIllustrationsFn({ data: { groupId, server: resolveGamedataServer(server) } }),
        staleTime: 60 * 60 * 1000,
        gcTime: 24 * 60 * 60 * 1000,
    });
}

/**
 * One group's ARCHIVE, or `null` when the running backend does not serve the
 * route (404). Null is the same first-class answer the illustrations query
 * gives, and for the same reason: the route shipped after the library did, so
 * a backend that predates it must yield a sheet with no archive segment rather
 * than a page that throws.
 *
 * A group that simply kept no archive is NOT a 404: the backend answers 200
 * with an empty `sections`, and 81 of the 87 EN groups look like that.
 */
export const getStoryArchiveFn = createServerFn({ method: "GET" })
    .inputValidator((data: { groupId: string; server: string }) => data)
    .handler(async ({ data: { groupId, server } }) => {
        const res = await backendFetch(gamedataPath(server, `/story/group/${encodeURIComponent(groupId)}/archive`));
        if (res.status === 404) return null;
        if (!res.ok) throw new Error(`Failed to load archive: ${res.status}`);
        return (await res.json()) as StoryArchive;
    });

export function storyArchiveQueryOptions(groupId: string, server: string = DEFAULT_GAMEDATA_SERVER) {
    return queryOptions({
        queryKey: ["story", "archive", resolveGamedataServer(server), groupId],
        queryFn: () => getStoryArchiveFn({ data: { groupId, server: resolveGamedataServer(server) } }),
        staleTime: 60 * 60 * 1000,
        gcTime: 24 * 60 * 60 * 1000,
    });
}

/**
 * What the COMMUNITY has read, or `null` when the running backend does not
 * serve the route (404). It is the same first-class null the archive and the
 * illustrations queries give, and it is not hypothetical here: the binary on
 * :3060 while this shipped answers 404, so the tab must render its own empty
 * state rather than throw a query error at the page.
 *
 * The backend caches the aggregate for 6 hours, so a shorter staleTime here
 * would only re-fetch a document that cannot have changed.
 */
export const getStoryCommunityFn = createServerFn({ method: "GET" })
    .inputValidator((server: string) => server)
    .handler(async ({ data: server }) => {
        const res = await backendFetch(gamedataPath(server, "/story/community"));
        if (res.status === 404) return null;
        if (!res.ok) throw new Error(`Failed to load community reading: ${res.status}`);
        return (await res.json()) as StoryCommunity;
    });

export function storyCommunityQueryOptions(server: string = DEFAULT_GAMEDATA_SERVER) {
    return queryOptions({
        queryKey: ["story", "community", resolveGamedataServer(server)],
        queryFn: () => getStoryCommunityFn({ data: resolveGamedataServer(server) }),
        staleTime: 6 * 60 * 60 * 1000,
        gcTime: 24 * 60 * 60 * 1000,
    });
}

File: backend/src/app/routes/story.rs
//! `GET /story/index`, `GET /story/{story_id}`,
//! `GET /story/group/{group_id}/illustrations`,
//! `GET /story/group/{group_id}/archive` and `GET /story/community`: the
//! Archives library, one parsed script, one group's art, one group's archive
//! and what the community has read. See `docs/story-reader.md` for the wire
//! contract.

use axum::{
    Json,
    extract::{Path, State},
};

use crate::app::services::story::{
    StoryArchive, StoryIllustrations, StoryIndex, get_group_archive, get_group_illustrations,
    get_story, get_story_index,
};
use crate::app::services::story_community::{self, StoryCommunity};
use crate::app::{error::ApiError, state::AppState};
use crate::core::hypergryph::constants::Server;
use crate::core::story::StoryScript;

/// The Archives library on the default server: every story group with its
/// stories, and every operator's records. `hasScript` says whether
/// `GET /story/{id}` will answer with a script.
#[utoipa::path(
    get,
    path = "/story/index",
    operation_id = "story_index",
    tag = "gamedata",
    responses(
        (status = 200, description = "The story library.", body = StoryIndex),
        (status = 429, response = crate::app::openapi::responses::RateLimited),
        (status = 500, response = crate::app::openapi::responses::InternalError),
        (status = 503, response = crate::app::openapi::responses::ServiceUnavailable)
    )
)]
pub async fn index(State(state): State<AppState>) -> Result<Json<StoryIndex>, ApiError> {
    Ok(Json(get_story_index(&state, state.default_server).await?))
}

/// `GET /{server}/story/index` - the story library from `{server}`. The
/// `/{server}` form reads that server's game data; the bare form reads the
/// default server.
#[utoipa::path(
    get,
    path = "/{server}/story/index",
    operation_id = "story_index_srv",
    tag = "gamedata",
    params(
        ("server" = String, Path, description = "Game server: `en`, `jp`, `kr`, `cn` or `tw`.")
    ),
    responses(
        (status = 200, description = "The story library.", body = StoryIndex),
        (status = 404, response = crate::app::openapi::responses::NotFound),
        (status = 429, response = crate::app::openapi::responses::RateLimited),
        (status = 500, response = crate::app::openapi::responses::InternalError),
        (status = 503, response = crate::app::openapi::responses::ServiceUnavailable)
    )
)]
pub async fn index_srv(
    State(state): State<AppState>,
    Path(server): Path<Server>,
) -> Result<Json<StoryIndex>, ApiError> {
    Ok(Json(get_story_index(&state, server).await?))
}

/// One story on the default server: its script as a generic command stream
/// and every asset it references, resolved to `/api/assets` paths. 404 when
/// the id is unknown, and 404 with a message when the id is in the library
/// but no script file backs it.
#[utoipa::path(
    get,
    path = "/story/{story_id}",
    operation_id = "story_detail",
    tag = "gamedata",
    params(
        ("story_id" = String, Path, description = "Story id from the library (`1stact_level_a001_01_beg`).")
    ),
    responses(
        (status = 200, description = "The parsed script.", body = StoryScript),
        (status = 404, response = crate::app::openapi::responses::NotFound),
        (status = 429, response = crate::app::openapi::responses::RateLimited),
        (status = 500, response = crate::app::openapi::responses::InternalError),
        (status = 503, response = crate::app::openapi::responses::ServiceUnavailable)
    )
)]
pub async fn detail(
    State(state): State<AppState>,
    Path(story_id): Path<String>,
) -> Result<Json<StoryScript>, ApiError> {
    get_story(&state, state.default_server, &story_id)
        .await?
        .map(Json)
        .ok_or(ApiError::NotFound)
}

/// `GET /{server}/story/{story_id}` - one story from `{server}`. The
/// `/{server}` form reads that server's game data; the bare form reads the
/// default server.
#[utoipa::path(
    get,
    path = "/{server}/story/{story_id}",
    operation_id = "story_detail_srv",
    tag = "gamedata",
    params(
        ("server" = String, Path, description = "Game server: `en`, `jp`, `kr`, `cn` or `tw`."),
        ("story_id" = String, Path, description = "Story id from the library (`1stact_level_a001_01_beg`).")
    ),
    responses(
        (status = 200, description = "The parsed script.", body = StoryScript),
        (status = 404, response = crate::app::openapi::responses::NotFound),
        (status = 429, response = crate::app::openapi::responses::RateLimited),
        (status = 500, response = crate::app::openapi::responses::InternalError),
        (status = 503, response = crate::app::openapi::responses::ServiceUnavailable)
    )
)]
pub async fn detail_srv(
    State(state): State<AppState>,
    Path((server, story_id)): Path<(Server, String)>,
) -> Result<Json<StoryScript>, ApiError> {
    get_story(&state, server, &story_id)
        .await?
        .map(Json)
        .ok_or(ApiError::NotFound)
}

/// Every background, CG and character sprite a group's stories reference, on
/// the default server. `group_id` is a `StoryGroup.id` (`main_0`, `act11side`,
/// `story_kroos_set_1`) or an operator `charId` (`char_124_kroos`) for that
/// operator's records: one route rather than two, because an operator's
/// records are split over one or two story sets and have no single group id,
/// while the two id spaces cannot collide (every operator id starts with
/// `char_`, no group id does). Answered from the warmed index, so no script
/// is read. A name with no file on disk is listed with `url: null`.
#[utoipa::path(
    get,
    path = "/story/group/{group_id}/illustrations",
    operation_id = "story_group_illustrations",
    tag = "gamedata",
    params(
        ("group_id" = String, Path, description = "Story group id or an operator `charId`.")
    ),
    responses(
        (status = 200, description = "The group's illustrations.", body = StoryIllustrations),
        (status = 404, response = crate::app::openapi::responses::NotFound),
        (status = 429, response = crate::app::openapi::responses::RateLimited),
        (status = 500, response = crate::app::openapi::responses::InternalError),
        (status = 503, response = crate::app::openapi::responses::ServiceUnavailable)
    )
)]
pub async fn illustrations(
    State(state): State<AppState>,
    Path(group_id): Path<String>,
) -> Result<Json<StoryIllustrations>, ApiError> {
    get_group_illustrations(&state, state.default_server, &group_id)
        .await
        .map(Json)
}

/// `GET /{server}/story/group/{group_id}/illustrations` - the same from
/// `{server}`. The `/{server}` form reads that server's game data; the bare
/// form reads the default server.
#[utoipa::path(
    get,
    path = "/{server}/story/group/{group_id}/illustrations",
    operation_id = "story_group_illustrations_srv",
    tag = "gamedata",
    params(
        ("server" = String, Path, description = "Game server: `en`, `jp`, `kr`, `cn` or `tw`."),
        ("group_id" = String, Path, description = "Story group id or an operator `charId`.")
    ),
    responses(
        (status = 200, description = "The group's illustrations.", body = StoryIllustrations),
        (status = 404, response = crate::app::openapi::responses::NotFound),
        (status = 429, response = crate::app::openapi::responses::RateLimited),
        (status = 500, response = crate::app::openapi::responses::InternalError),
        (status = 503, response = crate::app::openapi::responses::ServiceUnavailable)
    )
)]
pub async fn illustrations_srv(
    State(state): State<AppState>,
    Path((server, group_id)): Path<(Server, String)>,
) -> Result<Json<StoryIllustrations>, ApiError> {
    get_group_illustrations(&state, server, &group_id)
        .await
        .map(Json)
}

/// The game's own "from the archive" screen for one group: its logs,
/// landmarks, news, files, gallery, soundtrack and recordings, in the order
/// the game lists them, with every picture and clip resolved to an
/// `/api/assets` path. Only the sections the group HAS are listed, so a group
/// the archive table does not carry answers 200 with an empty `sections`
/// rather than a 404; an id the library does not list at all is a 404, the
/// same as the illustrations route. Answered from the warmed index, so no
/// table is re-read.
#[utoipa::path(
    get,
    path = "/story/group/{group_id}/archive",
    operation_id = "story_group_archive",
    tag = "gamedata",
    params(
        ("group_id" = String, Path, description = "Story group id or an operator `charId`.")
    ),
    responses(
        (status = 200, description = "The group's archive.", body = StoryArchive),
        (status = 404, response = crate::app::openapi::responses::NotFound),
        (status = 429, response = crate::app::openapi::responses::RateLimited),
        (status = 500, response = crate::app::openapi::responses::InternalError),
        (status = 503, response = crate::app::openapi::responses::ServiceUnavailable)
    )
)]
pub async fn archive(
    State(state): State<AppState>,
    Path(group_id): Path<String>,
) -> Result<Json<StoryArchive>, ApiError> {
    get_group_archive(&state, state.default_server, &group_id)
        .await
        .map(Json)
}

/// `GET /{server}/story/group/{group_id}/archive` - the same from `{server}`.
/// The `/{server}` form reads that server's game data; the bare form reads
/// the default server.
#[utoipa::path(
    get,
    path = "/{server}/story/group/{group_id}/archive",
    operation_id = "story_group_archive_srv",
    tag = "gamedata",
    params(
        ("server" = String, Path, description = "Game server: `en`, `jp`, `kr`, `cn` or `tw`."),
        ("group_id" = String, Path, description = "Story group id or an operator `charId`.")
    ),
    responses(
        (status = 200, description = "The group's archive.", body = StoryArchive),
        (status = 404, response = crate::app::openapi::responses::NotFound),
        (status = 429, response = crate::app::openapi::responses::RateLimited),
        (status = 500, response = crate::app::openapi::responses::InternalError),
        (status = 503, response = crate::app::openapi::responses::ServiceUnavailable)
    )
)]
pub async fn archive_srv(
    State(state): State<AppState>,
    Path((server, group_id)): Path<(Server, String)>,
) -> Result<Json<StoryArchive>, ApiError> {
    get_group_archive(&state, server, &group_id).await.map(Json)
}

/// What the community has READ, on the default server: how many accounts have
/// opened each story, and how far into each group they get.
///
/// Anonymous throughout: every field is a count over accounts and nothing
/// names one. An account is counted only while it has turned neither
/// `publicProfile` nor `shareStats` off. The aggregate is computed off the
/// request path and recomputed at most every six hours, so `computedAt` is how
/// old the numbers are rather than when they were asked for.
#[utoipa::path(
    get,
    path = "/story/community",
    operation_id = "story_community",
    tag = "gamedata",
    responses(
        (status = 200, description = "The community reading aggregate.", body = StoryCommunity),
        (status = 429, response = crate::app::openapi::responses::RateLimited),
        (status = 500, response = crate::app::openapi::responses::InternalError),
        (status = 503, response = crate::app::openapi::responses::ServiceUnavailable)
    )
)]
pub async fn community(State(state): State<AppState>) -> Result<Json<StoryCommunity>, ApiError> {
    Ok(Json(
        (*story_community::cached(&state, state.default_server).await?).clone(),
    ))
}

/// `GET /{server}/story/community` - the same from `{server}`. The `/{server}`
/// form reads that server's game data and its accounts; the bare form reads
/// the default server.
#[utoipa::path(
    get,
    path = "/{server}/story/community",
    operation_id = "story_community_srv",
    tag = "gamedata",
    params(
        ("server" = String, Path, description = "Game server: `en`, `jp`, `kr`, `cn` or `tw`.")
    ),
    responses(
        (status = 200, description = "The community reading aggregate.", body = StoryCommunity),
        (status = 404, response = crate::app::openapi::responses::NotFound),
        (status = 429, response = crate::app::openapi::responses::RateLimited),
        (status = 500, response = crate::app::openapi::responses::InternalError),
        (status = 503, response = crate::app::openapi::responses::ServiceUnavailable)
    )
)]
pub async fn community_srv(
    State(state): State<AppState>,
    Path(server): Path<Server>,
) -> Result<Json<StoryCommunity>, ApiError> {
    Ok(Json(
        (*story_community::cached(&state, server).await?).clone(),
    ))
}
